src/base/locked_queue_index.h

```diff
@@ -19,46 +26,108 @@ class LockedQueue<T>::Index {
   ui64 Size() const THREAD_UNSAFE {
     ui64 index_size = 0u;
     for (const auto& shard : index_) {
-      index_size += shard.size();
+      index_size += shard.tasks.size();
     }
     return index_size;
   }
 
-  void Put(QueueIterator it, ui32 shard) {
-    // TODO(ilezhankin): describe rationale for this place.
-    if (shard >= index_.size()) {
-      index_.resize(shard + 1);
+  void Put(QueueIterator it, const ui32 shard) {
+    EnsureShardExists(shard);
+    reverse_index_[&*it] = std::make_pair(
+        index_[shard].tasks.insert(index_[shard].tasks.end(), it),
+        shard);
+  }
+
+  bool ShardIsEmpty(const ui32 shard) THREAD_UNSAFE {
+    CHECK(shard < index_.size());
+    return index_[shard].tasks.empty();
+  }
+
+  void WaitForShard(const ui32 shard, UniqueLock& lock,
+                    const Seconds& timeout) THREAD_UNSAFE {
+    CHECK(shard < index_.size());
+    index_[shard].pop_condition_.wait_for(lock, timeout);
+  }
+
+  void NotifyShard(const ui32 shard) THREAD_UNSAFE {
```

Comment:
You should review all your `THREAD_UNSAFE` attributes, since at least this method is used in thread-safe `Push()` without locks.